### PR TITLE
703 Handle Workstep Execution Failures

### PR DESCRIPTION
**Pull request by sf316**
# Description

I'm currently in the process of implementing the mechanism by following the [official NestJS document](https://docs.nestjs.com/recipes/cqrs).

The current error message payload being used is:
```
{
      errorCode: 'xxx',
      errorMessage: 'Execution fails
}
```

## Related Issue

#703

## How Has This Been Tested

In progress.

## Screenshots (if appropriate)
In progress.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Request to be added as a Code Owner/Maintainer

## Checklist

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.
- [ ] I commit to abide by the Responsibilities of Code Owners/Maintainers.

### CODE REVIEW

File: examples/bri-3/src/bri/vsm/capabilites/handleWorkstepFailures/handleWorkstepFailuresCommand.handler.ts [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+        BpiMessageType.Err,
+      );
+
+      const initiatorChannel =
+        tx.fromBpiSubjectAccount.ownerBpiSubject.publicKey;
+      await this.messagingAgent.publishMessage(
+        initiatorChannel,
+        JSON.stringify(errorBpiMessage),
+      );
+      tx.updateStatusToAborted();
```
Comment (ognjenkurtic) [inline examples/bri-3/src/bri/vsm/capabilites/handleWorkstepFailures/handleWorkstepFailuresCommand.handler.ts:41]: We want to keep this and the line bellow in the original command handler, to make sure it is handled under a single db transaction later. I guess we do not need the HandleWorkstepFailuresCommand and handler at all and can instead log the issue, constrict the bpi message and publish it to the initiator channel in the WorkstepExecutionFailuresEventHandler

File: examples/bri-3/src/bri/vsm/capabilites/handleWorkstepFailuresEvents/workstepExecutionFailures.event.ts [hunk 2]
```diff
@@ -0,0 +1,3 @@
+export class WorkstepExecutionFailuresEvent {
+  constructor(public readonly id: string, public readonly err: string) {}
```
Comment (ognjenkurtic) [inline examples/bri-3/src/bri/vsm/capabilites/handleWorkstepFailuresEvents/workstepExecutionFailures.event.ts:2]: add here all relevant data necessary to log the error and construct a bpi message that will be sent to the initiator. No need to read the db again in the event handler. Please look at the other comment as well.

File: examples/bri-3/src/bri/vsm/capabilites/handleWorkstepFailures/handleWorkstepFailuresCommand.handler.ts (hunk 1 above)
Comment (sf316) [reply examples/bri-3/src/bri/vsm/capabilites/handleWorkstepFailures/handleWorkstepFailuresCommand.handler.ts]: @ognjenkurtic I have made adjustments according to your comments. Right now, I think saga is not necessary in this PR. Would you like me to remove that part as well?

File: examples/bri-3/src/bri/vsm/capabilites/executeVsmCycle/executeVsmCycleCommand.handler.ts [hunk 3]
```diff
@@ -29,8 +29,18 @@ export class ExecuteVsmCycleCommandHandler
       await this.txStorageAgent.updateTransactionStatus(tx);
 
       if (!this.agent.validateTransactionForExecution(tx)) {
-        tx.updateStatusToInvalid();
-        await this.txStorageAgent.updateTransactionStatus(tx);
+        this.eventBus.publish(
+          new WorkstepExecutionFailuresEvent(
+            tx.id,
```
Comment (ognjenkurtic) [inline examples/bri-3/src/bri/vsm/capabilites/executeVsmCycle/executeVsmCycleCommand.handler.ts:34]: I think you can pass the full tx as a param and then reference the required fields in the event handler from the tx object. This  way we reduce the number of params this ctor takes.

File: examples/bri-3/src/bri/vsm/capabilites/handleWorkstepFailuresEvents/workstepExecutionFailures.handler.ts [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+  implements IEventHandler<WorkstepExecutionFailuresEvent>
+{
+  constructor(
+    private readonly logger: LoggingService,
+    private readonly messagingAgent: MessagingAgent,
+  ) {}
+
+  handle(event: WorkstepExecutionFailuresEvent) {
+    this.logger.logError(
+      `Invalid transaction for execution with id ${event.id}: ${event.err}`,
```
Comment (ognjenkurtic) [inline examples/bri-3/src/bri/vsm/capabilites/handleWorkstepFailuresEvents/workstepExecutionFailures.handler.ts:20]: 'Failed execution of transaction with id {}. Error: {}'

File: examples/bri-3/src/bri/vsm/capabilites/handleWorkstepFailuresEvents/workstepExecutionFailures.handler.ts [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+    );
+
+    const errPayload = {
+      errorCode: 'xxx',
+      errorMessage: 'Execution fails',
```
Comment (ognjenkurtic) [inline examples/bri-3/src/bri/vsm/capabilites/handleWorkstepFailuresEvents/workstepExecutionFailures.handler.ts:25]: errorMessage  should be the same as the one which is logged for now

File: examples/bri-3/src/bri/vsm/capabilites/sagas/vsmFailures.sagas.ts [hunk 6]
```diff
@@ -0,0 +1,12 @@
+import { Injectable } from '@nestjs/common';
+import { ICommand, ofType, Saga } from '@nestjs/cqrs';
+import { Observable } from 'rxjs';
+import { WorkstepExecutionFailuresEvent } from '../handleWorkstepFailuresEvents/workstepExecutionFailures.event';
+
+@Injectable()
+export class VsmFailureSagas {
```
Comment (ognjenkurtic) [inline examples/bri-3/src/bri/vsm/capabilites/sagas/vsmFailures.sagas.ts:7]: yes, we can remote the saga for now.

File: examples/bri-3/src/bri/vsm/capabilites/handleWorkstepFailures/handleWorkstepFailuresCommand.handler.ts (hunk 1 above)
Comment (ognjenkurtic) [reply examples/bri-3/src/bri/vsm/capabilites/handleWorkstepFailures/handleWorkstepFailuresCommand.handler.ts]: Thanks, looks great. left a couple of smaller comments and it is ready to merge AFAIC. You can remote the saga.

Comment (sf316) [reply examples/bri-3/src/bri/vsm/capabilites/handleWorkstepFailures/handleWorkstepFailuresCommand.handler.ts]: @ognjenkurtic Done. Please review it. Thank you!

File: examples/bri-3/src/bri/vsm/capabilites/executeVsmCycle/executeVsmCycleCommand.handler.ts [hunk 7]
```diff
@@ -29,8 +29,11 @@ export class ExecuteVsmCycleCommandHandler
       await this.txStorageAgent.updateTransactionStatus(tx);
 
       if (!this.agent.validateTransactionForExecution(tx)) {
-        tx.updateStatusToInvalid();
-        await this.txStorageAgent.updateTransactionStatus(tx);
+        this.eventBus.publish(
+          new WorkstepExecutionFailuresEvent(tx, 'Validation Error'),
+        );
+        tx.updateStatusToAborted();
```
Comment (ognjenkurtic) [inline examples/bri-3/src/bri/vsm/capabilites/executeVsmCycle/executeVsmCycleCommand.handler.ts:35]: tx.updateStatusToInvalid();

Comment (sf316) [reply examples/bri-3/src/bri/vsm/capabilites/executeVsmCycle/executeVsmCycleCommand.handler.ts]: @ognjenkurtic Fixed. Apologizes for not reading the code carefully.

**Review by ognjenkurtic — APPROVED** [review]

File: examples/bri-3/src/bri/vsm/capabilites/executeVsmCycle/executeVsmCycleCommand.handler.ts (hunk 7 above)
Comment (ognjenkurtic) [reply examples/bri-3/src/bri/vsm/capabilites/executeVsmCycle/executeVsmCycleCommand.handler.ts]: no worries, approved now, looks great

**Review by Kasshern — APPROVED** [review]
Very nice work. clean code. Finally nestjs events in the repo.

**Pull request merged**